pkgs/yaml_edit/lib/src/utils.dart

```diff
@@ -277,6 +280,308 @@ String getLineEnding(String yaml) {
+    /// https://yaml.org/spec/1.2.2/#822-block-mappings
+    '-' || '?' || ':' => (compactCharOffset: startOffset, lineEndingIndex: -1),
+    _ => (compactCharOffset: -1, lineEndingIndex: -1)
+  };
+}
+
+typedef NextBlockNodeInfo = ({int nearestLineEnding, int nextNodeColStart});
+typedef BlockNodeOffset = ({int start, int end});
+
+/// Removes a block entry and its line ending from a [blockCollection] using the
+/// [nodeToRemoveOffset] provided. Any trailing comments are also removed.
+///
+/// If [blockCollection] is a [YamlMap], the chunk removed corresponds to the
+/// key-value pair represented by the offset. If [blockCollection] is a
+/// [YamlList], the chunk represents a single element within the list.
+///
+/// [nextBlockNodeInfo] is only called if the [blockCollection] has at least
+/// 2 entries and the entry being removed is not the last entry in the
+/// collection.
+SourceEdit removeBlockCollectionEntry(
+  String yaml, {
+  required YamlNode blockCollection,
+  required bool isFirstEntry,
+  required bool isSingleEntry,
+  required bool isLastEntry,
+  required BlockNodeOffset nodeToRemoveOffset,
+  required String lineEnding,
+  required NextBlockNodeInfo Function() nextBlockNodeInfo,
+}) {
+  final isBlockList = blockCollection is YamlList;
+
+  assert(
+    isBlockList || blockCollection is YamlMap,
+    'Expected a block map/list',
+  );
+
+  var makeNextNodeCompact = false;
+  var (:start, :end) = nodeToRemoveOffset;
+
+  if (start != 0 && !isSingleEntry) {
+    /// Try making it compact in case the collection's parent is a:
+    ///   - block sequence
+    ///   - explicit key
+    ///   - explicit key's explicit value.
+    if (isFirstEntry) {
+      final (:compactCharOffset, :lineEndingIndex) = indexOfCompactChar(
+        yaml,
+        start,
+      );
+
+      if (compactCharOffset != -1) {
+        start = compactCharOffset + 2; // Skip separation space.
+        makeNextNodeCompact = true;
+      } else {
+        start = lineEndingIndex + 1;
+      }
+    } else {
+      /// If not possible, just consume this node's indent. This prevents this
+      /// node from interfering with the next node.
+      start = yaml.lastIndexOf('\n', start) + 1;
+    }
+  }
+
+  var replacement = '';
+
+  // Skip empty lines to the last line break
+  end = indexOfLastLineEnding(yaml, yaml.indexOf('\n', end - 1));
+  end = min(++end, yaml.length); // Mark it for removal
+
+  if (isSingleEntry) {
+    /// Take back the last line ending even if multiple were skipped. This node
+    /// is sandwiched between other elements not in this map.
+    if (end < yaml.length - 1) {
+      end -= lineEnding == '\r\n' ? 2 : 1;
+    }
+
+    replacement = isBlockList ? '[]' : '{}';
+  } else if (!isLastEntry) {
+    final (:nearestLineEnding, :nextNodeColStart) = nextBlockNodeInfo();
+    final trueEndOffset = end - 1;
+
+    /// Make compact only if we are pointing to same line break from different
```

Comment:
nit. technically these are not documentation comment, just comments so `//` would be more accurate than `///`.

Not that this is important.